ni_measurementlink_service/session_management.py

```diff
@@ -138,6 +138,39 @@ def unreserve(self):
         """Unreserve sessions."""
         self._session_manager._unreserve_sessions(self._session_info)
 
+    def get_session_and_channel_for_pin(
+        self, pin: str, site: Optional[int] = None
+    ) -> Tuple[int, List[str]]:
+        """Returns the session information based on the given pin names."""
+        session_and_channel_info = self.get_sessions_and_channels_for_pins(pins=[pin], site=site)
+
+        if len(session_and_channel_info) != 1:
+            raise ValueError(f"Unsupported number of sessions: {len(self.session_info)}")
+        return session_and_channel_info[0]
+
+    def get_sessions_and_channels_for_pins(
+        self, pins: Union[str, List[str]], site: Optional[int] = None
+    ) -> List[Tuple[int, List[str]]]:
+        """Returns the session information based on the given pin names."""
+        pin_names = [pins] if isinstance(pins, str) else pins
+        session_and_channel_info = []
+        for session_index in range(len(self.session_info)):
+            channel_list = []
+            for pin in pin_names:
+                for channel_mapping in self.session_info[session_index].channel_mappings:
```

Comment:
I don't like iterating over channel_mappings multiple times because it may be large. I think it's better to iterate over channel_mappings once and match each element against pin_names.

Also, consider converting pin_names to a set for quicker matching.

Use a list comprehension:

```
channel_list = [
    mapping.channel
    for mapping in channel_mappings
    if mapping.pin_or_relay_name in pin_names and (site is None or mapping.site == site)
]
```

Reply:
Updated as per suggestions